File: code/language/src/main/java/org/fbme/ide/iec61499/repository/PlatformElementsOwner.java
package org.fbme.ide.iec61499.repository;

import org.fbme.lib.common.Element;
import org.fbme.lib.iec61499.IEC61499Factory;
import org.fbme.lib.st.STFactory;
import org.jetbrains.mps.openapi.model.SNode;

import java.util.HashMap;
import java.util.Map;

public class PlatformElementsOwner {
    private final Map<SNode, Element> myElements = new HashMap<>();
    private final PlatformElementAdapter myAdapter = MpsBridge.createElementAdapter(this);

    public <T> T getAdapter(SNode node, Class<T> requiredClass) {
        if (node == null) {
            return null;
        }
        Object adapter = myElements.computeIfAbsent(node, this::adapt);
        if (adapter == null) {
            myElements.remove(node);
        }
        if (requiredClass.isInstance(adapter)) {
            //noinspection unchecked
            return (T) adapter;
        }
        return null;
    }

    private Element adapt(SNode node) {
        return myAdapter.adapt(node);
    }

    public void dispose() {
        myElements.clear();
    }

    public IEC61499Factory getIEC61499Factory() {
        return MpsBridge.createIec61499Factory(this);
    }

    public STFactory getSTFactory() {
        return MpsBridge.createStFactory(this);
    }

}

File: code/language/src/main/kotlin/org/fbme/ide/iec61499/repository/PlatformElementsOwner.kt
package org.fbme.ide.iec61499.repository

import org.fbme.lib.common.Element
import org.fbme.lib.iec61499.IEC61499Factory
import org.fbme.lib.st.STFactory
import org.jetbrains.mps.openapi.model.SNode
import kotlin.reflect.KClass

open class PlatformElementsOwner {
    private val elements: MutableMap<SNode, Element?> = HashMap()
    private val adapter: PlatformElementAdapter = MpsBridge.createElementAdapter(this)

    fun getAdapterRaw(node: SNode): Element? {
        val adapter = elements.computeIfAbsent(node) { adapt(it) }
        if (adapter == null) elements.remove(node)
        return adapter
    }

    fun <T> getAdapterNullable(node: SNode?, requiredClass: Class<T>): T? =
        node?.let { getAdapterRaw(it) }?.let { requiredClass.cast(it) }

    fun <T> getAdapter(node: SNode, requiredClass: Class<T>): T =
        requiredClass.cast(getAdapterRaw(node))

    inline fun <reified T : Element> adapter(node: SNode) = getAdapterRaw(node) as T
    inline fun <reified T : Element> adapterOrNull(node: SNode) = getAdapterRaw(node) as? T

    private fun adapt(node: SNode): Element? {
        return adapter.adapt(node)
    }

    fun dispose() {
        elements.clear()
    }

    val iec61499Factory: IEC61499Factory
        get() = MpsBridge.createIec61499Factory(this)
    val stFactory: STFactory
        get() = MpsBridge.createStFactory(this)
}
